content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,209 @@ Like a segment tree on values rather than indices.
 	</Resource>
 </Resources>
 
-### Solution - Range K-th Smallest
+Suppose you want to support the following queries:
+
+- Given a range l, r count number of occurrences of value x.
+- Given a range l, r find the k smallest element
+
+With wavelet tree you can easily support those queries in log(n) time.
+
+<Info title="Additional information">
+  It is possible to support with updates in $\mathcal{O}(\log n^2)$
+</Info>
+
+## Wavelet Tree Structure
+
+A wavelet tree is a binary tree where each node represents a range of values.
+The root node covers the entire range, and each subsequent level splits the
+range into two halves.
+
+## First lets define our Data Structures
+
+We are going to maintain a segment tree over values instead of indices.
+Each segment contains indices whose values lie within the segment's range.
+Let's save those indicies in a vector.
+Notice that a index can be in at most log(N) segments
+
+<Spoiler title="Wavelet tree Visualization">
+
+Lets say our array is: [3,5,3,1,2,2,3,4,5,5]
+
+![Wavelet tree Visualization](https://imgur.com/Z7fpTFy.png)
```

Comment:
you can upload the image straight to usaco guide

Reply:
Will be done.

Reply:
when you're done you can resolve these convos by pressing the "Resolve Conversation" button